matrix_client/user.py

```diff
@@ -25,13 +25,24 @@ def __init__(self, api, user_id, displayname=None):
         self.displayname = displayname
         self.api = api
 
-    def get_display_name(self):
-        """ Get this users display name.
-            See also get_friendly_name()
+    def get_display_name(self, room=None):
+        """Get this user's display name.
+
+        See also get_friendly_name().
+
+        Args:
+            room (Room): Optional. If absent, return the global displayname for the user.
+                Otherwise, try return the display name of the user in the room, if any,
+                and default to the global displayname.
 
         Returns:
             str: Display Name
         """
+        if room:
+            try:
+                return room.members_displaynames[self.user_id]
+            except KeyError:
+                pass
```

Comment:
If `KeyError` here, we should raise the same error that gets raised if `self.api.get_display_name` fails. If a user doesn't have a displayname on their `m.room.member`, we shouldn't give a displayname in the context of a particular room.

Would be great if you could write a quick test-case for this particular scenario.

Reply:
Yep, defaulting to the global display name seems like a bad idea after all.

`self.api.get_display_name` doesn't raise anything and just returns `None` (which sure is unexpected according to the docstring of `User.get_display_name`), so here the try/except clause would just be replaced by `return room.members_displaynames.get(self.user_id)`.

However, it seems like `get_display_name` should indeed raise an exception, while if someone doesn't want that they should call `get_friendly_name` which ideally wraps around `get_display_name`, returning the user id if an exception was raised.

While we're at it, I don't think this is a good design. The spec doesn't say anything that suggests having two methods (to me it says that only `get_friendly_name` should exist, and should be called `get_display_name`). Also I can't see any use of having an exception raised in this kind of method.

So my proposal would be:
  - `User.get_display_name()` gives the global display name, or the user id if not set.
  - `User.get_display_name(room)` gives the display name corresponding to the room, or the user id if not set.
  - `User.get_friendly_name()` is deprecated and returns `User.get_display_name()`.

How's all that?

Reply:
Sounds great to me!